File: app/src/main/java/com/example/david/lists/widget/configview/WidgetConfigAdapter.java
package com.example.david.lists.widget.configview;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.example.david.lists.data.datamodel.UserList;
import com.example.david.lists.databinding.WidgetConfigListItemBinding;
import com.example.david.lists.view.userlistlist.UserListDiffCallback;

import java.util.ArrayList;
import java.util.List;

public final class WidgetConfigAdapter extends ListAdapter<UserList, WidgetConfigAdapter.WidgetConfigViewHolder>
        implements IWidgetConfigContract.Adapter {

    private final List<UserList> userLists;

    private final IWidgetConfigContract.ViewModel viewModel;

    public WidgetConfigAdapter(IWidgetConfigContract.ViewModel viewModel) {
        super(new UserListDiffCallback());
        this.userLists = new ArrayList<>();
        this.viewModel = viewModel;
    }

    @NonNull
    @Override
    public WidgetConfigViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new WidgetConfigViewHolder(
                WidgetConfigListItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false)
        );
    }

    @Override
    public void onBindViewHolder(@NonNull WidgetConfigViewHolder holder, int position) {
        holder.bindView(getItem(position));
    }

    @Override
    public void submitList(@Nullable List<UserList> list) {
        super.submitList(list);
        // In case the exact same List is submitted twice
        if (this.userLists != list) {
            userLists.clear();
            userLists.addAll(list);
        }
    }


    final class WidgetConfigViewHolder extends RecyclerView.ViewHolder
            implements View.OnClickListener {

        private final WidgetConfigListItemBinding binding;

        WidgetConfigViewHolder(WidgetConfigListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
            binding.getRoot().setOnClickListener(this);
        }

        private void bindView(UserList userList) {
            binding.tvTitle.setText(userList.getTitle());
        }

        @Override
        public void onClick(View v) {
            viewModel.userListClicked(
                    userLists.get(getAdapterPosition())
            );
        }
    }
}

File: app/src/main/java/com/example/david/lists/widget/configview/WidgetConfigAdapter.kt
package com.example.david.lists.widget.configview

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.david.lists.R
import com.example.david.lists.view.userlistlist.UserListDiffCallback
import com.example.domain.datamodel.UserList
import kotlinx.android.extensions.LayoutContainer
import kotlinx.android.synthetic.main.widget_config_list_item.*

class WidgetConfigAdapter(private val logic: IWidgetConfigContract.Logic) :
        ListAdapter<UserList, WidgetConfigAdapter.WidgetConfigViewHolder>(UserListDiffCallback()),
        IWidgetConfigContract.Adapter {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = WidgetConfigViewHolder(
            LayoutInflater.from(parent.context).inflate(R.layout.widget_config_list_item, parent, false)
    )

    override fun onBindViewHolder(holder: WidgetConfigViewHolder, position: Int) {
        holder.bindView(getItem(position).title)
    }

    override fun setData(list: List<UserList>) {
        super.submitList(list)
    }


    inner class WidgetConfigViewHolder(private val view: View) :
            RecyclerView.ViewHolder(view),
            LayoutContainer,
            View.OnClickListener {

        override val containerView: View?
            get() = view

        init {
            view.setOnClickListener(this)
        }

        fun bindView(title: String) {
            tv_title.text = title
        }

        override fun onClick(v: View) {
            logic.selectedUserList(adapterPosition)
        }
    }
}